pkg/pillar/cmd/volumemgr/updatestatus.go

```diff
@@ -269,38 +267,115 @@ func doUpdateContentTree(ctx *volumemgrContext, status *types.ContentTreeStatus)
 				status.SetErrorWithSource(err.Error(), types.ContentTreeStatus{}, time.Now())
 				return changed, false
 			}
-			// order is important
-			blobStatuses = append(blobs, blobStatuses...)
+			// if we have any, append them
+			if len(blobs) > 0 {
+				publishBlobStatus(ctx, blobs...)
+				// order is important; prepend to existing
+				blobStatuses = append(blobs, blobStatuses...)
+
+				// we changed it, so update the ContentTreeStatus
+				blobHashes := []string{}
+				for _, b := range blobStatuses {
+					blobHashes = append(blobHashes, b.Sha256)
+				}
+				status.Blobs = blobHashes
+				// Adding a blob to ContentTreeStatus and incrementing the refcount of that blob should be atomic as
+				// we would depend on that while we remove a blob from ContentTreeStatus and decrement
+				// the RefCount of that blob. In case if the blobs in a ContentTreeStatus in not in sync with the
+				// corresponding Blob's Refcount, then that would lead to Fatal error.
+				// If the same sha appears in multiple places in the ContentTree we intentionally add it twice to the list of
+				// Blobs so that we can have two reference counts on that blob.
+				// Add for the new ones
+				for _, b := range blobs {
+					AddRefToBlobStatus(ctx, b)
```

Comment:
@deitch I added this since otherwise it would Fatal due to the RefCount being zero when it tried to decrement it.

Reply:
Thanks @eriknordmark